contracts/core/EntryPoint.sol

```diff
@@ -581,42 +581,73 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
      * @param opIndex                            - The operation index.
      * @param op                                 - The user operation.
      * @param opInfo                             - The operation info.
-     * @param requiredPreFund                    - The required prefund amount.
      * @return context                           - The Paymaster-provided value to be passed to the 'postOp' function later
      * @return validationData                    - The Paymaster's validationData.
      */
     function _validatePaymasterPrepayment(
         uint256 opIndex,
         PackedUserOperation calldata op,
-        UserOpInfo memory opInfo,
-        uint256 requiredPreFund
+        UserOpInfo memory opInfo
     ) internal virtual returns (bytes memory context, uint256 validationData) {
         unchecked {
             uint256 preGas = gasleft();
             MemoryUserOp memory mUserOp = opInfo.mUserOp;
             address paymaster = mUserOp.paymaster;
+            uint256 requiredPreFund = opInfo.prefund;
             if (!_tryDecrementDeposit(paymaster, requiredPreFund)) {
                 revert FailedOp(opIndex, "AA31 paymaster deposit too low");
             }
             uint256 pmVerificationGasLimit = mUserOp.paymasterVerificationGasLimit;
-            try
-                IPaymaster(paymaster).validatePaymasterUserOp{gas: pmVerificationGasLimit}(
-                    op,
-                    opInfo.userOpHash,
-                    requiredPreFund
-                )
-            returns (bytes memory _context, uint256 _validationData) {
-                context = _context;
-                validationData = _validationData;
-            } catch {
-                revert FailedOpWithRevert(opIndex, "AA33 reverted", Exec.getReturnData(REVERT_REASON_MAX_LEN));
-            }
+            (context, validationData) = _callValidatePaymasterUserOp(opIndex, op, opInfo);
             if (preGas - gasleft() > pmVerificationGasLimit) {
                 revert FailedOp(opIndex, "AA36 over paymasterVerificationGasLimit");
             }
         }
     }
 
+    function _callValidatePaymasterUserOp(
+        uint256 opIndex,
+        PackedUserOperation calldata op,
+        UserOpInfo memory opInfo
+    ) internal returns (bytes memory context, uint256 validationData)  {
+        uint256 freePtr = _getFreePtr();
+        bytes memory validatePaymasterCall = abi.encodeCall(
+            IPaymaster.validatePaymasterUserOp,
+            (op, opInfo.userOpHash, opInfo.prefund)
+        );
+        address paymaster = opInfo.mUserOp.paymaster;
+        bool success;
+        assembly ("memory-safe") {
+            //call and return 3 first words: offset, validation, context-length
+            success := call(gas(), paymaster, 0, add(validatePaymasterCall, 0x20), mload(validatePaymasterCall), freePtr, 96)
```

Comment:
I think that if you read the entire return data here instead of limiting yourself to 96 bytes, you will only lose 64 bytes of "garbage data" per UserOp . However, it will remove the need to manually check and read the data with `returndatacopy` later .
What do you think?

Reply:
I read the first 3 words: offset, validationData, context-length
in a separte returndatacopy I read the entire context.
yes, I could read entire blob in a single returndatacopy , and then return a pointer into the middle of this struct..
that is:
always do:
```
 returndatacopy(freePtr, returndatasize())
context := add(freePtr,64)
```
(and then all sanity is the same)

is it much simpler?